File: python/transpose/transpose.py
import re
def transpose(lines):
    lines = lines.split("\n")
    num_lines = len(lines)
    anything_left = True
    char_index = 0
    groups = []
    while anything_left:
        anything_left = False
        out = ''
        for i in range(num_lines):
            cur_line = lines[i]
            if len(cur_line) > char_index:
                char = cur_line[char_index]
                out += '|' if char == ' ' else char
                if len(cur_line) > char_index + 1: anything_left = True
            else:
                out += ' '
        out = re.sub(r"\s+$", "", out)
        out = re.sub(r"\|", " ", out)
        groups.append(out)
        char_index += 1
    return "\n".join(groups)

File: ruby/transpose/transpose.rb
# null char as a placeholder for spaces
NPC = '\x00'
RNPC = Regexp.escape(NPC)

class Transpose
  def self.transpose(input)
    return '' if input == ''
    lines = input.split("\n").map { |line| line.split('') }
    max = lines.map(&:size).max
    out = Array.new(max, '')
    lines.each { |line|
      (0...max).each { |idx|
        out[idx] += (line[idx] || NPC)
      }
    }
    return out.map { |line|
      line.sub(/[#{RNPC}]*$/,'').gsub(/#{RNPC}/,' ')
    }.join("\n")
  end
end
